java/ql/src/experimental/Security/CWE/CWE-489/ServletMain.ql

```diff
@@ -0,0 +1,49 @@
+/**
+ * @name Main Method in Servlet
+ * @description Jave EE web applications with a main method.
+ * @kind problem
+ * @id java/main-method-in-servlet
+ * @tags security
+ *       external/cwe-489
+ */
+
+import java
+import semmle.code.java.frameworks.Servlets
+
+/** The java type `javax.servlet.Filter` */
+class ServletFilterClass extends Class {
+  ServletFilterClass() { this.getASupertype*().hasQualifiedName("javax.servlet", "Filter") }
+}
+
+/** Listener class in the package `javax.servlet` and `javax.servlet.http` */
+class ServletListenerClass extends Class {
+  // Various listener classes of Java EE such as ServletContextListener. They all have a name ending with the word "Listener".
+  ServletListenerClass() {
+    this.getASupertype*()
+        .getQualifiedName()
+        .regexpMatch([
+            "javax\\.servlet\\.[a-zA-Z]+Listener", "javax\\.servlet\\.http\\.[a-zA-Z]+Listener"
+          ])
+  }
+}
+
+/** The `main` method in `Servlet`. */
+class ServletMainMethod extends Method {
+  ServletMainMethod() {
+    (
+      this.getDeclaringType() instanceof ServletClass or
+      this.getDeclaringType() instanceof ServletFilterClass or
+      this.getDeclaringType() instanceof ServletListenerClass
+    ) and
+    this.hasName("main") and
+    this.isStatic() and
+    this.getReturnType() instanceof VoidType and
+    this.isPublic() and
+    this.getNumberOfParameters() = 1 and
+    this.getParameter(0).getType() instanceof Array and
+    not this.getDeclaringType().getName().matches("%Test%") // Simple check to exclude test classes to reduce FPs
```

Comment:
Can you add a similar clause to check if the Class's path contains `/test/`

Reply:
Thanks @pwntester for reviewing this PR. I've made the requested change.

Reply:
I think that is also useful but I refered to `getLocation().getFile().getAbsolutePath()...`

Reply:
Java packages matches file directory structure so they shall work the same. Please let me know you think this one is better if it's more intuitive.

Reply:
Maven test directory is not included in the package name. Eg: https://lgtm.com/projects/g/jhy/jsoup/snapshot/1d614593866749d4194dd8dea130a9242eb64930/files/src/test/java/org/jsoup/integration/servlets/Deflateservlet.java?sort=name&dir=ASC&mode=heatmap

You can keep your current check for packages for cases like:
https://lgtm.com/projects/g/AH7/testj/snapshot/a5f0665d8bfc1959cd2368b18bf04c696a7d589f/files/src/testcases/CWE113_HTTP_Response_Splitting/s01/CWE113_HTTP_Response_Splitting__Environment_addCookieServlet_01.java

and add a check for file paths containing `src/test/java` which should match maven test directory structure

Reply:
Thanks @pwntester for the explanation and the sample projects. I've made the change.